lib/ResultAggregator.ts

```diff
@@ -31,83 +31,101 @@ export class ResultAggregator implements IResultAggregator {
+          aggregate.timeMin = Math.min(aggregate.timeMin, result.time);
+
+          // If we haven't registered hash, we do so for full query result
+          if (aggregate.hash.length === 0) {
+            aggregate.hash = result.hash;
+          } else if (aggregate.hash !== result.hash) {
             inconsistentResults = true;
             aggregate.failures++;
           }
-          for (const [ index, timestamp ] of result.timestamps.entries()) {
-            if (timestampDivisors.length > index) {
-              timestampDivisors[index] += 1;
-              aggregate.timestamps[index] += timestamp;
-              aggregate.timestampsMax[index] = Math.max(aggregate.timestampsMax[index], timestamp);
-              aggregate.timestampsMin[index] = Math.min(aggregate.timestampsMin[index], timestamp);
-            } else {
-              timestampDivisors.push(1);
-              aggregate.timestamps.push(timestamp);
-              aggregate.timestampsMax.push(timestamp);
-              aggregate.timestampsMin.push(timestamp);
-            }
-          }
+        }
+        timestampsAll.push(result.timestamps);
+        if (result.timestamps.length > maxNumTimestamp) {
+          maxNumTimestamp = result.timestamps.length;
         }
       }
       if (inconsistentResults && !aggregate.error) {
         aggregate.error = new Error('Result hash inconsistency');
       }
-      if (successfulExecutions > 0) {
-        aggregate.time /= successfulExecutions;
-        aggregate.results /= successfulExecutions;
-        for (const [ index, timestampDivisor ] of timestampDivisors.entries()) {
-          aggregate.timestamps[index] /= timestampDivisor;
+
+      if (timestampsAll.length > 0) {
+        if (successfulExecutions > 0) {
+          aggregate.time /= successfulExecutions;
+          aggregate.results /= successfulExecutions;
         }
+
+        const processedTs = this.averageTimeStamps(timestampsAll, maxNumTimestamp);
+        aggregate.timestamps = processedTs.averageTs;
+        aggregate.timestampsMin = processedTs.minTs;
+        aggregate.timestampsMax = processedTs.maxTs;
       }
+
+      // Convert all possible leftover infinity / -infinity back to 0 for backward compatibility
+      aggregate.resultsMin = Number.isFinite(aggregate.resultsMin) ? aggregate.resultsMin : 0;
+      aggregate.resultsMax = Number.isFinite(aggregate.resultsMax) ? aggregate.resultsMax : 0;
+      aggregate.timeMin = Number.isFinite(aggregate.timeMin) ? aggregate.timeMin : 0;
+      aggregate.timeMax = Number.isFinite(aggregate.timeMax) ? aggregate.timeMax : 0;
+
       aggregates.push(aggregate);
     }
     return aggregates;
   }
 
+  public averageTimeStamps(timestampsAll: number[][], maxNumTimestamps: number): IProcessedTimestamps {
+    const totalTs: number[] = <number[]> Array.from({ length: maxNumTimestamps }).fill(0);
+    const maxTs: number[] = <number[]> Array.from({ length: maxNumTimestamps }).fill(-Number.POSITIVE_INFINITY);
```

Comment:
I think `Number.NEGATIVE_INFINITY` would work here, as well (instead of `-Number.Positive_INFINITY`)?

Reply:
Agreed